consensus/src/consensus/mod.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use rand::{thread_rng, Rng};
+use snarkos_storage::{DynStorage, SerialBlock, SerialTransaction, VMRecord};
+use snarkvm_algorithms::CRH;
+use snarkvm_dpc::{
+    testnet1::{
+        instantiated::{Components, InstantiatedDPC, SerialNumberNonce},
+        payload::Payload,
+        PublicParameters,
+    },
+    Account,
+    AccountScheme,
+    DPCComponents,
+};
+use tokio::sync::{mpsc, oneshot};
+
+use crate::{error::ConsensusError, Address, ConsensusParameters, DynLedger, MemoryPool};
+
+use self::message::{ConsensusMessage, TransactionResponse};
+use anyhow::*;
+
+mod inner;
+pub use inner::ConsensusInner;
+mod message;
+pub use message::*;
+
+pub struct Consensus {
+    pub parameters: ConsensusParameters,
+    pub public_parameters: PublicParameters<Components>,
+    genesis_block: SerialBlock,
+    sender: mpsc::Sender<ConsensusMessage>,
+}
+
+impl Consensus {
+    pub fn new(
+        parameters: ConsensusParameters,
+        public_parameters: PublicParameters<Components>,
+        genesis_block: SerialBlock,
+        ledger: DynLedger,
+        storage: DynStorage,
+        memory_pool: MemoryPool,
+    ) -> Arc<Self> {
+        let (sender, receiver) = mpsc::channel(256);
+        let created = Arc::new(Self {
+            parameters,
+            public_parameters,
+            genesis_block,
+            sender,
+        });
+
+        let created2 = created.clone();
+        tokio::spawn(async move {
+            ConsensusInner {
```

Comment:
The name `ConsensusInner` is IMO a bit unclear, especially since it contains a `Consensus` object inside of it; the `public` member also sounds unclear to me.

Reply:
its a relatively common pattern -- we have a private subtype (not in OO sense) which can access its public supertype. private subtype is running an agent/it's own thread, while public supertype is a read-only arc-copyable dispatch handle with some (in this case) extra data tagged (parameters)

Reply:
I don't have strong feelings about this, but in that case it would be a good idea to document this in some general design document so that it's more clear, as most of the code is fairly pattern-agnostic in terms of member names.